src/model/differ.js

```diff
@@ -0,0 +1,774 @@
+	 * @param {Map} newAttributes Map, map iterator or compatible object that contains attributes after change.
+	 * @returns {Array.<Object>} Array containing one or more diff items.
+	 */
+	_getAttributesDiff( range, oldAttributes, newAttributes ) {
+		// Results holder.
+		const diffs = [];
+
+		// Clone new attributes as we will be performing changes on this object.
+		newAttributes = new Map( newAttributes );
+
+		// Look through old attributes.
+		for ( const [ key, oldValue ] of oldAttributes ) {
+			// Check what is the new value of the attribute (or if it was removed).
+			const newValue = newAttributes.has( key ) ? newAttributes.get( key ) : null;
+
+			// If values are different (or attribute was removed)...
+			if ( newValue !== oldValue ) {
+				// Add diff item.
+				diffs.push( {
+					type: 'attribute',
+					position: range.start,
+					range: Range.createFromRange( range ),
+					length: 1,
+					attributeKey: key,
+					attributeOldValue: oldValue,
+					attributeNewValue: newValue,
+					changeCount: this._changeCount++
+				} );
+
+				// Prevent returning two diff items for the same change.
+				newAttributes.delete( key );
+			}
+		}
+
+		// Look through new attributes that weren't handled above.
+		for ( const [ key, newValue ] of newAttributes ) {
+			// Each of them is a new attribute. Add diff item.
+			diffs.push( {
+				type: 'attribute',
+				position: range.start,
+				range: Range.createFromRange( range ),
+				length: 1,
+				attributeKey: key,
+				attributeOldValue: null,
+				attributeNewValue: newValue,
+				changeCount: this._changeCount++
+			} );
+		}
+
+		return diffs;
+	}
+}
+
+// Returns an array that is a copy of passed child list with the exception that text nodes are split to one or more
+// objects, each representing one character and attributes set on that character.
+function _getChildrenSnapshot( children ) {
+	const snapshot = [];
+
+	for ( const child of children ) {
+		if ( child.is( 'text' ) ) {
+			for ( let i = 0; i < child.data.length; i++ ) {
+				snapshot.push( {
+					name: '$text',
+					attributes: new Map( child.getAttributes() )
+				} );
+			}
+		} else {
+			snapshot.push( {
+				name: child.name,
+				attributes: new Map( child.getAttributes() )
+			} );
+		}
+	}
+
+	return snapshot;
+}
+
+// Generates array of actions for given changes set.
+// It simulates what `diff` function does.
+// Generated actions are:
+// - 'e' for 'equal' - when item at that position did not change,
+// - 'i' for 'insert' - when item at that position was inserted,
+// - 'r' for 'remove' - when item at that position was removed,
+// - 'a' for 'attribute' - when item at that position has it attributes changed.
+function _generateActionsFromChanges( oldChildrenLength, changes ) {
+	const actions = [];
+
+	let offset = 0;
+	let oldChildrenHandled = 0;
+
+	for ( const change of changes ) {
+		if ( change.offset > offset ) {
+			actions.push( ...'e'.repeat( change.offset - offset ).split( '' ) );
```

Comment:
Comment "fill holes between end of one change and beginning of another with 'e'".